File: java-test/src/test/java/io/aerisconsulting/catadioptre/test/WithGenericTest.java
/*
 * Copyright 2021 AERIS-Consulting e.U.
 *
 * AERIS-Consulting e.U. licenses this file to you under the Apache License, version 2.0 (the
 * "License"); you may not use this file except in compliance with the License. You may obtain a
 * copy of the License at:
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package io.aerisconsulting.catadioptre.test;

import static org.mockito.Mockito.mock;

import java.util.HashMap;
import java.util.Map;
import org.junit.jupiter.api.Test;

@SuppressWarnings("java:S2699")
class WithGenericTest {

	WithGeneric instance = new WithGeneric();

	@Test
	void shouldCompileCallOfAFunctionWithGenericsAsArgumentAndReturnedValue() {
		Map<String, Converted> argument = new HashMap();
		argument.put("1", mock(Converted.class));

		TestableWithGeneric.filterMap(instance, argument);
	}

	@Test
	void shouldCompileCallOfAFunctionWithGenericStarAsArgument() {
		TestableWithGeneric.processGenericStarAndHasNoReturnedValue(instance, mock(Specification.class));
	}

	@Test
	void shouldCompileCallOfAFunctionWithGenericReferencingItselfAsArgument() {
		TestableWithGeneric.processGenericReferencingItself(instance, mock(Specification.class));
	}

}

File: kotlin-test/src/test/kotlin/io/aerisconsulting/catadioptre/test/WithGenericTest.kt
package io.aerisconsulting.catadioptre.test

import io.aerisconsulting.catadioptre.test.catadioptre.filterMap
import io.aerisconsulting.catadioptre.test.catadioptre.processGenericReferencingItself
import io.aerisconsulting.catadioptre.test.catadioptre.processGenericStar
import io.mockk.mockk
import org.junit.jupiter.api.Test

internal class WithGenericTest {

    @Test
    internal fun `should compile call of a function with generics as argument and returned value`() {
        val instance = WithGeneric()
        instance.filterMap(mapOf("1" to mockk<Converted>()))
    }

    @Test
    internal fun `should compile call of a function with generic star as argument`() {
        val instance = WithGeneric()
        instance.processGenericStar(mockk<Specification<Int, Any?, *>>())
    }

    @Test
    internal fun `should compile call of a function with generic referencing itself as argument`() {
        val instance = WithGeneric()
        instance.processGenericReferencingItself(mockk<Specification<Any?, Any?, out Specification<Any?, Any?, *>>>())
    }
}